File: leetcode/solved/43_Multiply_Strings/solution.cpp
/*
 * solution.cpp
 * Copyright (C) 2018 sungmin <smoh2044@gmail.com>
 *
 * Distributed under terms of the MIT license.
 */


#include <bits/stdc++.h>


using namespace std;


class Solution {
public:
    string sum(string num1, string num2){
        stringstream ss;
        int i = num1.size()-1;
        int j = num2.size()-1;
        int c = 0;
        while(i>=0 || j>=0){
            int n1 = i >= 0 ? num1[i]-48 : 0;
            int n2 = j >= 0 ? num2[j]-48 : 0;
            c += n1+n2;
            ss << to_string(c%10);
            c = c/10;
            i--;
            j--;
        }
        if(c != 0) ss << to_string(c);
        string s = ss.str();
        reverse(s.begin(), s.end());
        return s;
    }
    string multiply(string num1, string num2) {
        if(num1.compare("0") == 0 || num2.compare("0") == 0) return "0";
        stringstream ss;
        string s = "0";
        int c = 0;
        int size1 = num1.size()-1;
        for(int i=size1; i>=0; --i){
            int n1 = num1[i]-48;
            for(int t=0; t<size1-i; ++t){
                ss << '0';
            }
            for(int j=num2.size()-1; j>=0; --j){
                int n2 = num2[j]-48;
                c += n1*n2;
                ss << to_string(c%10);
                c /= 10;
            }
            if(c != 0) ss << to_string(c);
            c = 0;
            string s2 = ss.str();
            reverse(s2.begin(), s2.end());
            s = sum(s, s2);
            ss.str("");
        }
        if(c != 0){
            s = to_string(c) + s;
        }
        return s;
    }
};



string input(){
    string s;
    getline(cin, s);
    return s;
}


int main(){
    string s1 = input();
    string s2 = input();
    cout << Solution().multiply(s1, s2) << endl;

    return 0;
}

File: leetcode/solved/43_Multiply_Strings/solution.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2020 sungminoh <smoh2044@gmail.com>
#
# Distributed under terms of the MIT license.

"""
Given two non-negative integers num1 and num2 represented as strings, return the product of num1 and num2, also represented as a string.

Note: You must not use any built-in BigInteger library or convert the inputs to integer directly.

Example 1:
Input: num1 = "2", num2 = "3"
Output: "6"
Example 2:
Input: num1 = "123", num2 = "456"
Output: "56088"

Constraints:

	1 <= num1.length, num2.length <= 200
	num1 and num2 consist of digits only.
	Both num1 and num2 do not contain any leading zero, except the number 0 itself.
"""
import sys
import itertools
import pytest


class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        if num1 == '0' or num2 == '0':
            return '0'
        def mul(num, digit):
            d = int(digit)
            ret = ''
            c = 0
            for a in reversed(num):
                c, n = divmod(int(a)*d + c, 10)
                ret += str(n)
            if c:
                ret += str(c)
            return ret[::-1]

        def plus(num1, num2):
            ret = ''
            c = 0
            for a, b in itertools.zip_longest(reversed(num1), reversed(num2)):
                a = a or '0'
                b = b or '0'
                c, n = divmod(int(a)+int(b)+c, 10)
                ret += str(n)
            if c:
                ret += str(c)
            return ret[::-1]

        ret = '0'
        for i, d in enumerate(reversed(num2)):
            ret = plus(ret, mul(num1, d) + '0'*i)

        return ret


@pytest.mark.parametrize('num1, num2, expected', [
    ("2", "3", "6"),
    ("123", "456", "56088"),
])
def test(num1, num2, expected):
    assert expected == Solution().multiply(num1, num2)


if __name__ == '__main__':
    sys.exit(pytest.main(["-s", "-v"] + sys.argv))
